emacs/doom.d/config.org
#+TITLE: My Doom Emacs config
#+AUTHOR: Justine Kizhakkinedath
#+EMAIL: justine@kizhak.com
#+LANGUAGE: en
#+STARTUP: noinlineimages
#+PROPERTY: header-args:emacs-lisp :tangle yes :cache yes :results silent :padline no
#+OPTIONS: toc:nil
#+EXPORT_FILE_NAME: README.md

* My Doom Emacs config

#+MARKDOWN: [[_TOC_]]
* About Emacs

[[https://imgs.xkcd.com/comics/cautionary.png]]

Emacs changes how you /think/ about programming.

Emacs is *totally introspectable*. You can always find out 'what code runs
when I press this button?'.

Emacs is an *incremental programming environment*. There's no edit-compile-run
cycle.
There isn't even an edit-run cycle. You can execute snippets of code and
gradually turn them into a finished project. There's no distinction between
your editor and your interpreter.

Emacs is a *mutable environment*. You can set variables, tweak functions
with advice, or redefine entire functions. Nothing is off-limits.

Emacs provides *functionality without applications*. Rather than separate
applications, functionality is all integrated into your Emacs instance.
Amazingly, this works. Ever wanted to use the same snippet tool for writing
C++ classes as well as emails?

Emacs is full of *incredible software concepts that haven't hit the mainstream
yet*. For example:
- Many platforms have a single item clipboard. Emacs has an *infinite clipboard*.
- If you undo a change, and then continue editing, you can't redo the original
  change. Emacs allows *undoing to any historical state*, even allowing
  tree-based exploration of history.
- Emacs supports a *reverse variable search*: you can find variables with
  a given value.
- You can perform *structural editing* of code, allowing you to make changes
  without breaking syntax. This works for lisps (paredit) and non-lisps (smartparens).
- Many applications use a modal GUI: for example, you can't do other edits
  during a find-and-replace operation. Emacs provides *recursive editing* that
  allow you to suspend what you're currently doing, perform other edits, then
  continue the original task.
Emacs has a *documentation culture*. Emacs includes a usage manual, a lisp
programming manual, pervasive docstrings and even an interactive tutorial.

Emacs has a *broad ecosystem*. If you want to edit code in a niche language,
there's probably an Emacs package for it.

Emacs doesn't have a monopoly on good ideas, and there are other great tools
out there.
Nonetheless, we believe the Emacs learning curve pays off.

[[https://i.stack.imgur.com/7Cu9Z.jpg]]

/This beautifully written *About EMACS* section credits to [[https://github.com/remacs/remacs][Remacs]]./

Also if you want to read more about Emacs, checkout these articles
- [[https://two-wrongs.com/why-you-should-buy-into-the-emacs-platform][Why You Should Buy Into the Emacs Platform]]
- [[https://www.reddit.com/r/emacs/comments/hmtwbt/i_use_emacs_because/][I use emacs because...]]

** Editor comparison

[[https://imgs.xkcd.com/comics/real_programmers.png]]

Also this graph by[[https://github.com/tecosaur/emacs-config/blob/master/config.org][ tecosaur]] describes the best

[[./images/editor-comparison.png]]

[[https://tecosaur.github.io/emacs-config/config.html][Tecosaur's config]]
* About my config

This Emacs config is a work of many hours painful polishing and trying to create
a perfect work environment.

** Contribution

If you spotted a bug or you have any suggestions, please fill in an issue.
If you have something to fix, feel free to create a pull request.
** Installation

Use the install directions from [[https://github.com/hlissner/doom-emacs][Doom Emacs]].

For specific instructions use ~devfile.toml~.
** License

   You have the permission to use, modify, distribute in any way you want.

   However, what is /free/ stays /free/. After all, this is [[file:LICENSE.org][GPL]].

   All the files inside the ~doom.d~ is licensed under the terms of GNU GENERAL PUBLIC
   LICENSE Version 3.

   For more information [[file:LICENSE.org][read the license]].

** Repository mirroring

[[https://gitlab.com/justinekizhak/dotfiles/-/tree/master/emacs/doom.d][My Doom Emacs config on Gitlab]] (Main repo)

[[https://github.com/justinekizhak/dotfiles/tree/master/emacs/doom.d][My Doom Emacs config on Github]] (Mirror)

** Screenshots

*** Dashboard
[[./images/dashboard.png]]


*** Org mode

[[./images/org-mode.png]]

*** Python mode

[[./images/python-mode.png]]
** Some helper guide
*** [[https://noelwelsh.com/posts/2019-01-10-doom-emacs.html][Doom Emacs Workflows]]
** Sorting the headlines

To sort the headers in this file run =M-x org-sort-entries= and select alphabetically order.

** Special Thanks

Everyone starts somewhere, and I started here.

- [[https://github.com/MatthewZMD/.emacs.d][MatthewZMD]]
- [[https://github.com/hlissner/doom-emacs][Henrik Lissner's Doom Emacs]]

* Lexical Binding

Use lexical-binding. [[https://nullprogram.com/blog/2016/12/22/][Why?]]

#+BEGIN_QUOTE
Until Emacs 24.1 (June 2012), Elisp only had dynamically scoped variables,
a feature, mostly by accident, common to old lisp dialects. While dynamic
scope has some selective uses, it’s widely regarded as a mistake for local
variables, and virtually no other languages have adopted it.
#+END_QUOTE

#+BEGIN_SRC emacs-lisp
;;; config.el --- -*- lexical-binding: t -*-
#+END_SRC

* Personal Information
Let's set some variables with basic user information.

#+BEGIN_SRC emacs-lisp
(setq user-full-name "Justine Kizhakkinedath"
      user-mail-address "justine@kizhak.com")
#+END_SRC

* Defining constants

** Are we running on a GNU/Linux system?

#+BEGIN_SRC emacs-lisp
(defconst *sys/linux*
  (eq system-type 'gnu/linux))
#+END_SRC
** Are we running on a GUI Emacs?

#+BEGIN_SRC emacs-lisp
(defconst *sys/gui*
  (display-graphic-p))
#+END_SRC

** Are we running on a Mac system?

#+BEGIN_SRC emacs-lisp
(defconst *sys/mac*
  (eq system-type 'darwin))
#+END_SRC

** Are we running on a WinTel system?

#+BEGIN_SRC emacs-lisp
(defconst *sys/win32*
  (eq system-type 'windows-nt))
#+END_SRC

** Are you a ROOT user?

#+BEGIN_SRC emacs-lisp
(defconst *sys/root*
  (string-equal "root" (getenv "USER")))
#+END_SRC

** Do we have python3?

#+BEGIN_SRC emacs-lisp
(defconst *python3*
  (executable-find "python3"))
#+END_SRC

** Check basic requirements for EAF to run.

#+BEGIN_SRC emacs-lisp
(defconst *eaf-env*
  (and *sys/linux* *sys/gui* *python3*
       (executable-find "pip")
       (not (equal (shell-command-to-string "pip freeze | grep '^PyQt\\|PyQtWebEngine'") ""))))
#+END_SRC

** Do we have clangd?

#+BEGIN_SRC emacs-lisp
(defconst *clangd*
  (or (executable-find "clangd")  ;; usually
      (executable-find "/usr/local/opt/llvm/bin/clangd")))  ;; macOS
#+END_SRC

** Do we have gcc?

#+BEGIN_SRC emacs-lisp
(defconst *gcc*
  (executable-find "gcc"))
#+END_SRC

** Do we have git?

#+BEGIN_SRC emacs-lisp
(defconst *git*
  (executable-find "git"))
#+END_SRC

** Do we have Maven?

#+BEGIN_SRC emacs-lisp
(defconst *mvn*
  (executable-find "mvn"))
#+END_SRC

** Do we have pdflatex?

#+BEGIN_SRC emacs-lisp
(defconst *pdflatex*
  (executable-find "pdflatex"))
#+END_SRC

** Do we have python?

#+BEGIN_SRC emacs-lisp
(defconst *python*
  (executable-find "python"))
#+END_SRC

** Do we have ripgrep?

#+BEGIN_SRC emacs-lisp
(defconst *rg*
  (executable-find "rg"))
#+END_SRC

** Do we have tr?

#+BEGIN_SRC emacs-lisp
(defconst *tr*
  (executable-find "tr"))
#+END_SRC

* Some Emacs defaults

#+BEGIN_SRC emacs-lisp
(use-package emacs
  :preface
  (defvar ian/indent-width 2) ; change this value to your preferred width
  :config
  (setq
    ring-bell-function 'ignore       ; minimise distraction
    frame-resize-pixelwise t
    default-directory "~/")

  (tool-bar-mode -1)
  (menu-bar-mode -1)

  ;; increase line space for better readability
  (setq-default line-spacing 3)

  ;; Always use spaces for indentation
  (setq-default indent-tabs-mode nil
                tab-width ian/indent-width))

#+END_SRC

* Better editing experience

** Automatically refreshes the buffer for changes outside of Emacs

Auto refreshes every 2 seconds. Don’t forget to refresh the version control status as well.

#+BEGIN_SRC emacs-lisp
(use-package autorevert
  :defer t
  :ensure nil
  :config
  (global-auto-revert-mode +1)
  (setq auto-revert-interval 2
        auto-revert-check-vc-info t
        auto-revert-verbose nil))
#+END_SRC

** Confirm kill process

Don’t bother confirming killing processes

#+BEGIN_SRC emacs-lisp
(use-package files
  :defer t
  :config
  (setq confirm-kill-processes nil))
#+END_SRC

** Disable scroll bar

#+BEGIN_SRC emacs-lisp
(use-package scroll-bar
  :defer t
  :ensure nil
  :config (scroll-bar-mode -1))
#+END_SRC

** History

#+BEGIN_SRC emacs-lisp :tangle no
(use-package recentf
  :defer t
  :ensure nil
  :hook (after-init . recentf-mode)
  :custom
  (recentf-auto-cleanup "05:00am")
  (recentf-max-saved-items 200)
  (recentf-exclude '((expand-file-name package-user-dir)
                     ".cache"
                     ".cask"
                     ".elfeed"
                     "bookmarks"
                     "cache"
                     "ido.*"
                     "persp-confs"
                     "recentf"
                     "undo-tree-hist"
                     "url"
                     "COMMIT_EDITMSG\\'")))

;; When buffer is closed, saves the cursor location
(save-place-mode 1)

;; Set history-length longer
(setq-default history-length 500)
#+END_SRC

** Modernize selection behavior

Replace the active region just by typing text, just like modern editors

#+BEGIN_SRC emacs-lisp :tanlge no
(use-package delsel
  :disabled
  :ensure nil
  :config (delete-selection-mode +1))
#+END_SRC

#+BEGIN_SRC emacs-lisp
(setq delete-selection-mode t)
#+END_SRC

** Show matching parentheses

Reduce the highlight delay to instantly.

#+BEGIN_SRC emacs-lisp
(use-package paren
  :defer t
  :ensure nil
  :init (setq show-paren-delay 0.5)
  :config (show-paren-mode +1))
#+END_SRC

* Appearance

** Dashboard with images

Use the image in the dotfiles folder as the dashboard splash image

#+BEGIN_SRC emacs-lisp
(add-hook! '(+doom-dashboard-mode-hook)
           ;; Crypto logo
           (setq fancy-splash-image "~/dotfiles/emacs/doom.d/images/crypto.png"))
#+END_SRC

** Setting up some frame defaults

Maximize the frame by default on start-up. Set the font to BlexMono, if BlexMono is installed.

#+BEGIN_SRC emacs-lisp
(use-package frame
  :ensure t
  :config
  (add-to-list 'default-frame-alist '(fullscreen . maximized)))
#+END_SRC

* Custom Keybinding

** Window management

*** Jump to any window using =SPC w M-w=

Unbind existing =SPC w C-w= for Ace window

#+BEGIN_SRC emacs-lisp
(map! :leader
      (:prefix ("w")
        "C-w" nil))
#+END_SRC

Bind =SPC w M-w= for Ace window

#+BEGIN_SRC emacs-lisp
(map! :leader
      (:prefix ("w" . "window")
        :desc "Jump to any window using Ace" "M-w" #'ace-window))
#+END_SRC

** ⌘ + a → Select all

#+BEGIN_SRC emacs-lisp
(map! "M-a" #'mark-whole-buffer)
#+END_SRC

** ⌘ + s → Save file

#+BEGIN_SRC emacs-lisp
(map! "M-s" #'save-buffer)
#+END_SRC

** ⌘ + v → Paste menu

#+BEGIN_SRC emacs-lisp
(map! "M-v" #'counsel-yank-pop)
#+END_SRC

* Packages

# ** Ace-popup

# #+BEGIN_SRC emacs-lisp
# (use-package ace-popup-menu
#   :defer t)
# #+END_SRC

# ** Annotate

# [[https://github.com/bastibe/annotate.el][Github repo]]

# This package provides a minor mode annotate-mode, which can add annotations to arbitrary files without changing the files themselves. This is very useful for code reviews.

# #+BEGIN_SRC emacs-lisp :tangle no
# (use-package annotate
#   :defer t)
# #+END_SRC

** Apex Legends title

Use a random quote of a character from [[https://www.ea.com/games/apex-legends/play-now-for-free][Apex Legends]] as your frame title.

Requires [[https://pypi.org/project/apex-legends-voicelines/][Apex Legends voicelines]] python package.

*** Install the package using

#+BEGIN_SRC sh :tangle no
pipx install apex-legends-voicelines
#+END_SRC

*** Usage

#+BEGIN_SRC emacs-lisp
(setq frame-title-format (shell-command-to-string "apex-voicelines"))

(defun change-emacs-title-apex ()
  "Change your Emacs frame title using the voicelines of `Apex Legends' characters.
This command requires `apex-legends-voicelines' python package."
  (interactive)
  (setq frame-title-format (shell-command-to-string "apex-voicelines")))
#+END_SRC

# ** Atomic chrome

# Edit browser text in Emacs

# #+BEGIN_SRC emacs-lisp
# (use-package atomic-chrome
#   :defer t)
# (add-hook 'emacs-startup-hook (lambda ()
#                                   (atomic-chrome-start-server)))
# #+END_SRC

** Clipmon

#+BEGIN_SRC emacs-lisp
(add-to-list 'after-init-hook 'clipmon-mode-start)
#+END_SRC

** Dired

#+BEGIN_SRC emacs-lisp :tangle no
(use-package dired
  :defer t
  :ensure nil
  :bind
  (("C-x C-j" . dired-jump)
   ("C-x j" . dired-jump-other-window))
  :custom
  ;; Always delete and copy recursively
  (dired-recursive-deletes 'always)
  (dired-recursive-copies 'always)
  ;; Auto refresh Dired, but be quiet about it
  (global-auto-revert-non-file-buffers t)
  (auto-revert-verbose nil)
  ;; Quickly copy/move file in Dired
  (dired-dwim-target t)
  ;; Move files to trash when deleting
  (delete-by-moving-to-trash t)
  ;; Load the newest version of a file
  (load-prefer-newer t)
  ;; Detect external file changes and auto refresh file
  (auto-revert-use-notify nil)
  (auto-revert-interval 3) ; Auto revert every 3 sec
  :config
  ;; Enable global auto-revert
  (global-auto-revert-mode t)
  ;; Reuse same dired buffer, to prevent numerous buffers while navigating in dired
  (put 'dired-find-alternate-file 'disabled nil)
  :hook
  (dired-mode . (lambda ()
                  (dired-hide-details-mode)
                  (local-set-key (kbd "<mouse-2>") #'dired-find-alternate-file)
                  (local-set-key (kbd "RET") #'dired-find-alternate-file)
                  (local-set-key (kbd "^")
                                 (lambda () (interactive) (find-alternate-file ".."))))))
#+END_SRC

** Drag lines

*** Vertically

#+BEGIN_SRC emacs-lisp
(map!
    :n "M-k" #'drag-stuff-up    ; drags line up
    :n "M-j" #'drag-stuff-down)  ; drags line down
#+END_SRC

*** Horizontally

#+BEGIN_SRC emacs-lisp
(with-eval-after-load 'evil-org
  (map!
    :n "M-l" #'evil-org->       ; indents line to left
    :n "M-h" #'evil-org-<))      ; indents line to right
#+END_SRC

** Easy escape

Now no more double backslash hell. [[https://github.com/cpitclaudel/easy-escape][Github repo]]

[[https://raw.githubusercontent.com/cpitclaudel/easy-escape/master/img/easy-escape.png]]

#+BEGIN_SRC emacs-lisp
(use-package easy-escape
  :defer t
  :config
    (set-face-attribute 'easy-escape-face nil :foreground "red"))
#+END_SRC

** Evil snipe

#+BEGIN_SRC emacs-lisp
(use-package evil-snipe
  :defer t
  :config
  (setq evil-snipe-scope 'visible)
  (setq evil-snipe-repeat-scope 'buffer)
  (setq evil-snipe-spillover-scope 'whole-buffer))
#+END_SRC

# ** EWW

# Emacs Web Wowser, the HTML-based Emacs Web Browser.

# #+BEGIN_SRC emacs-lisp :tangle no
# (use-package eww
#   :defer t
#   :commands (eww)
#   :hook (eww-mode . (lambda ()
#                       "Rename EWW's buffer so sites open in new page."
#                       (rename-buffer "eww" t)))
#   :config
#   ;; I am using EAF-Browser instead of EWW
#   (unless *eaf-env*
#     (setq browse-url-browser-function 'eww-browse-url))) ; Hit & to browse url with system browser
# #+END_SRC
** Flycheck

Flycheck, a syntax checking extension.

#+BEGIN_SRC emacs-lisp :tangle no
(use-package flycheck
  :defer t
  :hook (prog-mode . flycheck-mode)
  :custom

CHANGELOG.md
# Changelog

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.10.0]
### Added
    - doom submodule.
    - doom init file.
    - spacemacs submodule.
### Update
    - install instructions.

## [0.9.1]
### Added
    - Contributing
    - Code of conduct
    - README.
### Update
    - spacemacs.el

## [0.9.0]
### Added
    - New framework Zim
### Update
    - install.yaml for zimfw.

## [0.8.0]
### Added
    - New facility to install header package from local repo.

## [0.7.0]
### Added
    - Issue templates

## [0.6.0]
### Added
    - tmux themepack and clean up.
### Update
    - zshrc.
    - tmux

## [0.5.0]
### Added
    - Script to install packages in separate venv.

## [0.4.0]
### Added
    - spacemacs app.
    - install files.
    - copy feature for tmux.
### Update
    - spacemacs.el.
    - scripts

## [0.3.0]
### Added
    - templates.

## [0.2.0]
### Added
    - font
    - vim
    - tmux
    - spacemacs.el
    - scripts folder
    - zsh
    - submodules
        - powerlevel9k
        - zsh-autosuggestions
        - zsh-sytax-hilighting

## [0.1.0]

[Unreleased]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.10.0...develop
[0.2.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.1.0...v0.2.0
[0.3.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.2.0...v0.3.0
[0.4.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.3.0...v0.4.0
[0.5.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.4.0...v0.5.0
[0.6.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.5.0...v0.6.0
[0.7.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.6.0...v0.7.0
[0.8.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.7.0...v0.8.0
[0.9.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.8.0...v0.9.0
[0.9.1]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.9.0...v0.9.1
[0.10.0]: https://gitlab.com/justinekizhak/dotfiles/compare/v0.9.1...v0.10.0


emacs/doom.d/config.el
;;; config.el --- -*- lexical-binding: t -*-
(setq user-full-name "Justine Kizhakkinedath"
      user-mail-address "justine@kizhak.com")
(defconst *sys/linux*
  (eq system-type 'gnu/linux))
(defconst *sys/gui*
  (display-graphic-p))
(defconst *sys/mac*
  (eq system-type 'darwin))
(defconst *sys/win32*
  (eq system-type 'windows-nt))
(defconst *sys/root*
  (string-equal "root" (getenv "USER")))
(defconst *python3*
  (executable-find "python3"))
(defconst *eaf-env*
  (and *sys/linux* *sys/gui* *python3*
       (executable-find "pip")
       (not (equal (shell-command-to-string "pip freeze | grep '^PyQt\\|PyQtWebEngine'") ""))))
(defconst *clangd*
  (or (executable-find "clangd")  ;; usually
      (executable-find "/usr/local/opt/llvm/bin/clangd")))  ;; macOS
(defconst *gcc*
  (executable-find "gcc"))
(defconst *git*
  (executable-find "git"))
(defconst *mvn*
  (executable-find "mvn"))
(defconst *pdflatex*
  (executable-find "pdflatex"))
(defconst *python*
  (executable-find "python"))
(defconst *rg*
  (executable-find "rg"))
(defconst *tr*
  (executable-find "tr"))
(use-package emacs
  :preface
  (defvar ian/indent-width 2) ; change this value to your preferred width
  :config
  (setq
    ring-bell-function 'ignore       ; minimise distraction
    frame-resize-pixelwise t
    default-directory "~/")

  (tool-bar-mode -1)
  (menu-bar-mode -1)

  ;; increase line space for better readability
  (setq-default line-spacing 3)

  ;; Always use spaces for indentation
  (setq-default indent-tabs-mode nil
                tab-width ian/indent-width))
(use-package autorevert
  :defer t
  :ensure nil
  :config
  (global-auto-revert-mode +1)
  (setq auto-revert-interval 2
        auto-revert-check-vc-info t
        auto-revert-verbose nil))
(use-package files
  :defer t
  :config
  (setq confirm-kill-processes nil))
(use-package scroll-bar
  :defer t
  :ensure nil
  :config (scroll-bar-mode -1))
(use-package delsel
  :disabled
  :ensure nil
  :config (delete-selection-mode +1))
(setq delete-selection-mode t)
(use-package paren
  :defer t
  :ensure nil
  :init (setq show-paren-delay 0.5)
  :config (show-paren-mode +1))
(add-hook! '(+doom-dashboard-mode-hook)
           ;; Crypto logo
           (setq fancy-splash-image "~/dotfiles/emacs/doom.d/images/crypto.png"))
(use-package frame
  :ensure t
  :config
  (add-to-list 'default-frame-alist '(fullscreen . maximized)))
(map! :leader
      (:prefix ("w")
        "C-w" nil))
(map! :leader
      (:prefix ("w" . "window")
        :desc "Jump to any window using Ace" "M-w" #'ace-window))
(map! "M-a" #'mark-whole-buffer)
(map! "M-s" #'save-buffer)
(map! "M-v" #'counsel-yank-pop)
(setq frame-title-format (shell-command-to-string "apex-voicelines"))

(defun change-emacs-title-apex ()
  "Change your Emacs frame title using the voicelines of `Apex Legends' characters.
This command requires `apex-legends-voicelines' python package."
  (interactive)
  (setq frame-title-format (shell-command-to-string "apex-voicelines")))
(add-to-list 'after-init-hook 'clipmon-mode-start)
(map!
    :n "M-k" #'drag-stuff-up    ; drags line up
    :n "M-j" #'drag-stuff-down)  ; drags line down
(with-eval-after-load 'evil-org
  (map!
    :n "M-l" #'evil-org->       ; indents line to left
    :n "M-h" #'evil-org-<))      ; indents line to right
(use-package easy-escape
  :defer t
  :config
    (set-face-attribute 'easy-escape-face nil :foreground "red"))
(use-package evil-snipe
  :defer t
  :config
  (setq evil-snipe-scope 'visible)
  (setq evil-snipe-repeat-scope 'buffer)
  (setq evil-snipe-spillover-scope 'whole-buffer))
(use-package goto-line-preview
  :defer t
  :config
    (global-set-key [remap goto-line] 'goto-line-preview))
(use-package hydra
  :defer t)
(use-package iedit
  :defer t)
(use-package indent-tools
  :defer t
  :after (hydra)
  :bind ("C-c >" . #'indent-tools-hydra/body))
(setq lsp-ui-doc-max-height 30)
(setq lsp-ui-doc-max-width 150)
(use-package org
  :defer t
  :config
  (setq org-startup-with-inline-images nil)
  (setq org-startup-shrink-all-tables t)
  (setq org-use-property-inheritance t)
  (setq org-hide-emphasis-markers t)
  ; Fix `org-cycle' bug
  (map! :map org-mode-map
        :n "<tab>" 'org-cycle)
  ; Add plantUML
  (add-to-list 'org-src-lang-modes '("plantuml" . plantuml))
  (setq org-plantuml-jar-path "~/plantuml.jar")
  (setq plantuml-default-exec-mode 'jar)
  ; Add graphviz
  (add-to-list 'org-src-lang-modes  '("dot" . graphviz-dot))
  (setq org-ellipsis "⬎"))
   ;; ➡, ⚡, ▼, ↴, , ∞, ⬎, ⤷, ⤵
(font-lock-add-keywords 'org-mode
                        '(("^ *\\([-]\\) "
                           (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

(let* ((variable-tuple
        (cond ((x-list-fonts "ETBembo")         '(:font "ETBembo"))
              ((x-list-fonts "Source Sans Pro") '(:font "Source Sans Pro"))
              ((x-list-fonts "Lucida Grande")   '(:font "Lucida Grande"))
              ((x-list-fonts "Verdana")         '(:font "Verdana"))
              ((x-family-fonts "Sans Serif")    '(:family "Sans Serif"))
              (nil (warn "Cannot find a Sans Serif Font.  Install Source Sans Pro."))))
       (headline           `(:inherit default :weight bold)))

  (custom-theme-set-faces
   'user
   `(org-level-8 ((t (,@headline ,@variable-tuple))))
   `(org-level-7 ((t (,@headline ,@variable-tuple))))
   `(org-level-6 ((t (,@headline ,@variable-tuple))))
   `(org-level-5 ((t (,@headline ,@variable-tuple :height 1.1))))
   `(org-level-4 ((t (,@headline ,@variable-tuple :height 1.2))))
   `(org-level-3 ((t (,@headline ,@variable-tuple :height 1.3))))
   `(org-level-2 ((t (,@headline ,@variable-tuple :height 1.4))))
   `(org-level-1 ((t (,@headline ,@variable-tuple :height 1.5))))
   `(org-document-title ((t (,@headline ,@variable-tuple :height 2.0 :underline nil))))))


(custom-theme-set-faces
 'user
 '(variable-pitch ((t (:family "ETBembo" :height 160))))
 '(fixed-pitch ((t ( :family "Fira Code" :height 160)))))

(add-hook 'org-mode-hook 'variable-pitch-mode)
(add-hook 'org-mode-hook 'visual-line-mode)

(custom-theme-set-faces
 'user
 '(org-block ((t (:inherit fixed-pitch))))
 '(org-code ((t (:inherit (shadow fixed-pitch)))))
 '(org-document-info ((t (:foreground "dark orange"))))
 '(org-document-info-keyword ((t (:inherit (shadow fixed-pitch)))))
 '(org-indent ((t (:inherit (org-hide fixed-pitch)))))
 '(org-link ((t (:foreground "royal blue" :underline t))))
 '(org-meta-line ((t (:inherit (font-lock-comment-face fixed-pitch)))))
 '(org-property-value ((t (:inherit fixed-pitch))) t)
 '(org-special-keyword ((t (:inherit (font-lock-comment-face fixed-pitch)))))
 '(org-table ((t (:inherit fixed-pitch :foreground "#83a598"))))
 '(org-tag ((t (:inherit (shadow fixed-pitch) :weight bold :height 0.8))))
 '(org-verbatim ((t (:inherit (shadow fixed-pitch))))))
(setq org-agenda-files (list "~/org/project/" "~/org/todo.org"))

(setq
  org-deadline-warning-days 7
  org-agenda-breadcrumbs-separator " ❱ "
  org-directory "~/org")
(customize-set-value
    'org-agenda-category-icon-alist
    `(
      ("work" "~/.doom.d/icons/money-bag.svg" nil nil :ascent center)
      ("chore" "~/.doom.d/icons/loop.svg" nil nil :ascent center)
      ("events" "~/.doom.d/icons/calendar.svg" nil nil :ascent center)
      ("todo" "~/.doom.d/icons/checklist.svg" nil nil :ascent center)
      ("walk" "~/.doom.d/icons/walk.svg" nil nil :ascent center)
      ("solution" "~/.doom.d/icons/solution.svg" nil nil :ascent center)))
(setq-hook! org-mode
  org-log-done t
  org-columns-default-format "%60ITEM(Task) %20TODO %10Effort(Effort){:} %10CLOCKSUM"
  org-global-properties (quote (("Effort_ALL" . "0:15 0:30 0:45 1:00 2:00 3:00 4:00 5:00 6:00 0:00")
                                ("STYLE_ALL" . "habit")))
  org-archive-location "~/org/archive/todo.org.gpg::")

(setq org-agenda-block-separator (string-to-char " "))

(setq org-agenda-custom-commands
      '(("o" "My Agenda"
         ((todo "TODO" ()
                      (org-agenda-overriding-header "\n⚡ Do Today:\n⎺⎺⎺⎺⎺⎺⎺⎺⎺")
                      (org-agenda-remove-tags t)
                      (org-agenda-prefix-format " %-2i %-15b")
                      (org-agenda-todo-keyword-format ""))

          (agenda "" (
                      (org-agenda-start-day "+0d")
                      (org-agenda-span 5)
                      (org-agenda-overriding-header "⚡ Schedule:\n⎺⎺⎺⎺⎺⎺⎺⎺⎺")
                      (org-agenda-repeating-timestamp-show-all nil)
                      (org-agenda-remove-tags t)
                      (org-agenda-prefix-format   "  %-3i  %-15b %t%s")
                      (org-agenda-todo-keyword-format " ☐ ")
                      (org-agenda-current-time-string "⮜┈┈┈┈┈┈┈ now")
                      (org-agenda-scheduled-leaders '("" ""))
                      (org-agenda-time-grid (quote ((daily today remove-match)
                                                    (0900 1200 1500 1800 2100)
                                                    "      " "┈┈┈┈┈┈┈┈┈┈┈┈┈")))))))))
(use-package org-bullets
  :defer t)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
(setq org-latex-hyperref-template "\\hypersetup{\n pdfauthor={%a},\n pdftitle={%t},\n pdfkeywords={%k},\n pdfsubject={%d},\n pdfcreator={%c}, \n pdflang={%L}, \n colorlinks = true}\n")
(use-package ox-reveal
    :defer t
    :config
    (setq org-reveal-root "/Users/justinkizhakkinedath/revealjs")
    (setq org-reveal-mathjax t))
(use-package ox-gfm
  :defer t)
(use-package org-ref
  :defer t
  :config
  (setq reftex-default-bibliography '("~/org/references.bib"))

  ;; see org-ref for use of these variables
  (setq org-ref-bibliography-notes "~/org/notes.org"
        org-ref-default-bibliography '("~/org/references.bib")
        org-ref-pdf-directory "~/org/bibtex-pdfs/"))

(setq bibtex-completion-bibliography "~/org/references.bib"
      bibtex-completion-library-path "~/org/bibtex-pdfs"
      bibtex-completion-notes-path "~/org/helm-bibtex-notes")

;; open pdf with system pdf viewer (works on mac)
(setq bibtex-completion-pdf-open-function
  (lambda (fpath)
    (start-process "open" "*open*" "open" fpath)))
(use-package parinfer
  :defer t)
(use-package powerthesaurus
  :defer t)
(map! :leader
      (:prefix ("a" . "applications")
        :desc "Use powerthesaurus to fetch better word" "p" #'powerthesaurus-lookup-word-dwim))
(use-package projectile
  :defer t
  :config
    (setq  projectile-project-search-path '("~/projects")))
(use-package deadgrep
  :defer t
  :config
    (map! :leader
      (:prefix ("a" . "applications")
        :desc "Open Ripgrep interface" "r" #'deadgrep)))
(use-package string-inflection
  :defer t)
(map! :leader
    (:prefix ("a" . "applications")
        :desc "Cycle through string case using String-inflection" "c" #'string-inflection-all-cycle))
(use-package treemacs-magit
  :defer t
  :after (treemacs magit))
(setq-default magit-process-password-prompt-regexps
  '("^\\(Enter \\)?[Pp]assphrase\\( for \\(RSA \\)?key '.*'\\)?: ?$"
    ;; Match-group 99 is used to identify the "user@host" part.
    "^\\(Enter \\)?[Pp]assword\\( for '\\(https?://\\)?\\(?99:.*\\)'\\)?: ?$"
    ;; Pinentry Curses box in the terminal when used with GnuPG
    "Please enter the passphrase for the ssh key"
    "^.*'s password: ?$"
    "^Yubikey for .*: ?$"
    "^Enter PIN for .*: ?$"))
(use-package undo-tree
  :defer t
  :custom
  (undo-tree-visualizer-diff t)
  (undo-tree-visualizer-timestamps t))
(add-hook 'vterm-mode-hook #'goto-address-mode)
(map! :map vterm-mode-map
      :n "P" #'vterm-yank
      :n "p" #'vterm-yank)
(use-package json-mode
  :defer t
  :mode "\\.json\\'")
;; Enable backup
(setq make-backup-files t)

;; Backup by copying
(setq backup-by-copying t)
(defun xah-save-all-unsaved ()
  "Save all unsaved files. no ask.
Version 2019-11-05"
  (interactive)
  (save-some-buffers t ))

;; when switching out of emacs, all unsaved files will be saved
(add-hook 'focus-out-hook 'xah-save-all-unsaved)
(defhydra hydra-paste (:color red
                       :hint nil)
  "\n[%s(length kill-ring-yank-pointer)/%s(length kill-ring)] \
 [_C-j_/_C-k_] cycles through yanked text, [_p_/_P_] pastes the same text \
 above or below. Anything else exits."
  ("C-j" evil-paste-pop)
  ("C-k" evil-paste-pop-next)
  ("p" evil-paste-after)
  ("P" evil-paste-before))

(map! :nv "p" #'hydra-paste/evil-paste-after
      :nv "P" #'hydra-paste/evil-paste-before)
(global-auto-revert-mode 1)
(setq mac-command-modifier 'meta)
